### PR TITLE
multi_disk_random_hotplug: Add codes to handle the hot_unplug error caused by cmd '__com.redhat_drive_del'

### PR DESCRIPTION
**Pull request by ngu-niny**
1. Add codes to handle unplug error caused by cmd '__com.redhat_drive_del'
2. Change all monitors to QMP ones
3. Add sleep time between/after hotplugs/unplugs

ID: 1303513

Signed-off-by: Nini Gu <ngu@redhat.com>

### CODE REVIEW

**Comment by ngu-niny** [comment]
Test PASS:

'1-smp_8.16384m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.multi_disk_random_hotplug.all_types' Finshed 2017-01-12 01:50:11 2017-01-12 01:57:35 PASS 444.0026
'2-smp_8.16384m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.multi_disk_random_hotplug.single_type' Finshed 2017-01-12 01:57:35 2017-01-12 01:59:26 PASS 111.17228
'3-smp_8.16384m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.multi_disk_random_hotplug.parallel.all_types' Finshed 2017-01-12 01:59:26 2017-01-12 02:01:16 PASS109.918462
'4-smp_8.16384m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.multi_disk_random_hotplug.parallel.single_type' Finshed 2017-01-12 02:01:16 2017-01-12 02:02:59 PASS 102.455768

**Comment by CongLi** [comment]
@ngu-niny 
Could you please check comment 2 in ID ?
seems "__com.redhat_drive_add" also has problem.

**Comment by ngu-niny** [comment]
@CongLi For  "__com.redhat_drive_add" issue in comment 2 of the bug, not clear about the issue since no log attached. So I will not handle it in this PR.

File: qemu/tests/multi_disk_random_hotplug.py [hunk 1]
```diff
@@ -179,7 +179,6 @@ def insert_into_qdev(qdev, param_matrix, no_disks, params, new_devices):
                 for dev in devs:
                     if dev in qdev:
                         qdev.remove(dev, recursive=True)
-                formats.remove(fmt)
```
Comment (xutian) [inline qemu/tests/multi_disk_random_hotplug.py]: it's very very useful, it use to reduce while loop, please don't drop it.

File: qemu/tests/multi_disk_random_hotplug.py [hunk 2]
```diff
@@ -273,6 +272,9 @@ def _unplug(new_devices, qdev, monitor, prefix=""):
         unplug_outs = []
         unplug_devs = []
         for device in new_devices[::-1]:    # unplug all devices
+            # No need to use cmd '__com.redhat_drive_del' to delete the disks
```
Comment (xutian) [inline qemu/tests/multi_disk_random_hotplug.py]: why not need ?

File: qemu/tests/multi_disk_random_hotplug.py (hunk 1 above)
Comment (ngu-niny) [inline qemu/tests/multi_disk_random_hotplug.py]: I have updated the code, now I add codes that only won't to remove the 'fmt' for qid existed issue in function insert_into_qdev()

File: qemu/tests/multi_disk_random_hotplug.py (hunk 2 above)
Comment (ngu-niny) [inline qemu/tests/multi_disk_random_hotplug.py]: @xutian That's for the parallel case, HMP has no command '__com.redhat_drive_del'; besides, currently, to upplug a drive, we don't need to use the command, only comd 'device_del' is enough.

**Comment by ngu-niny** [comment]
Have updated the code by handling unplug error caused by cmd '__com.redhat_drive_del' instead of deleting it directly. Many thanks @xutian for the help.

Test PASS:
'1-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.multi_disk_random_hotplug.all_types' Finshed 2017-02-19 11:30:47 2017-02-19 11:39:22 PASS 514.799017
'2-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.multi_disk_random_hotplug.single_type' Finshed 2017-02-19 11:39:22 2017-02-19 11:47:48 PASS 505.821497
'3-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.multi_disk_random_hotplug.parallel.all_types' Finshed 2017-02-19 11:47:48 2017-02-19 11:52:55 PASS 306.550194
'4-smp_4.8192m.repeat1.run_test.Host_RHEL.m7.u4.qcow2.virtio_scsi.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.multi_disk_random_hotplug.parallel.single_type' Finshed 2017-02-19 11:52:55 2017-02-19 11:57:56 PASS 301.35206

File: qemu/tests/multi_disk_random_hotplug.py [hunk 3]
```diff
@@ -276,7 +276,15 @@ def _unplug(new_devices, qdev, monitor, prefix=""):
             if device in qdev:  # Some devices are removed with previous one
                 time.sleep(unplug_sleep)
                 unplug_devs.append(device)
-                unplug_outs.append(device.unplug(monitor))
+                try:
+                    output = device.unplug(monitor)
+                except Exception:
+                    # Pass error caused by using cmd '__com.redhat_drive_del'
```
Comment (xutian) [inline qemu/tests/multi_disk_random_hotplug.py]: Please add comments like:
driver unplug after unplug block device by command 'device_del' in qemu >=2.6, but it not in old qemu. Below update is  compatible with the new version of qemu

Comment (ngu-niny) [inline qemu/tests/multi_disk_random_hotplug.py]: @xutian Commnet added, thanks!

File: qemu/tests/multi_disk_random_hotplug.py [hunk 4]
```diff
@@ -276,7 +276,20 @@ def _unplug(new_devices, qdev, monitor, prefix=""):
             if device in qdev:  # Some devices are removed with previous one
                 time.sleep(unplug_sleep)
                 unplug_devs.append(device)
-                unplug_outs.append(device.unplug(monitor))
+                try:
+                    output = device.unplug(monitor)
+                except Exception:
```
Comment (xutian) [inline qemu/tests/multi_disk_random_hotplug.py]: better to catch "qemu_monitor.MonitorError" here.

Comment (ngu-niny) [inline qemu/tests/multi_disk_random_hotplug.py]: Updated. Thanks.

**Comment by xutian** [comment]
Acked

**Comment by xutian** [comment]
Merged, thanks @ngu-niny and @CongLi

**Pull request merged**